### PR TITLE
Catalog Of My Things: Full Project

### PR DESCRIPTION
**Pull request by peterdtitan**
# CATALOG OF MY THINGS:

- [LIVE DEMO(VIDEO PRESENTATION)](https://drive.google.com/file/d/1a5sZ1Ca4rxdPMueHmSFK97FR7E3_90cE/view?usp=sharing)

#### Logic

- Each class from the UML class diagram was created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes were implemented.
- All properties visible in the diagram were defined and set up in the constructor method. _Exception: properties for the 1-to-many relationships were NOT set in the constructor method. Instead, they have a custom setter method created._
- All methods visible in the diagram were implemented:
    - add_item method (in all classes that have that method)
        - takes an instance of the Item class as an input.
        - adds the input item to the collection of items.
        - adds self as a property of the item object (by using the correct setter from the item object).
    - can_be_archived?() in the Item class
        - returns true if published_date is older than 10 years.
        - otherwise, it returns false.
    - move_to_archive() in the Item class
        - reuses can_be_archived?() method.
        - changes the archived property to true if the result of the can_be_archived?() method is true.
        - does nothing if the result of the can_be_archived?() method is false.
    - can_be_archived?() in the Book class
        - overrides the method from the parent class.
        - returns true if the parent's method returns true OR if cover_state equals "bad".
        - otherwise, it returns false.
    - can_be_archived?() in the MusicAlbum class
        - overrides the method from the parent class.
        - returns true if the parent's method returns true AND if on_spotify equals true.
        - otherwise, it returns false.
    - can_be_archived?() in the Movie class
        - overrides the method from the parent class.
        - returns true if the parent's method returns true OR if silent equals true.
        - otherwise, it returns false.
    - can_be_archived?() in the Game class
        - overrides the method from the parent class.
        - returns true if the parent's method returns true AND if last_played_at is older than 2 years.
        - otherwise, it returns false.
- Added unit tests for all implemented methods.

#### User interface

- Created a main.rb file that serves as the console app entry point.
- The console app, at the start:
    - Presents the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
    - Has a way to quit the app.
- The following options are available:
    - List all books
    - List all music albums
    - List all movies
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - List all sources (e.g. 'From a friend, 'Online shop')
    - Add a book
    - Add a music album
    - Add a movie
    - Add a game
 - All data is preserved by saving collections in .json files.

### Database
- Created a schema.sql file with tables that will be analogical to the structure of the classes in the app. As we cannot implement inheritance in the database tables - we added all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### CODE REVIEW

File: spec/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    it 'sets the cover state' do
+      expect(book_instance.cover_state).to eq(cover_state)
+    end
+  end
+
+  context '#id' do
+    it 'returns a unique identifier for the book' do
+      expect(book_instance.id).not_to be_nil
+    end
+  end
```
Comment (jerryowusu) [inline spec/book_spec.rb:27]: As part of the project requirements, all public methods in all classes should have unit tests. In view of this kindly provide unit test for the `can_be_archived` method in the book class. :+1:

File: item.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @id = id || Random.rand(101..1000)
+    @publish_date = publish_date
+    @archived = archived
+  end
+
+  def label=(label)
+    @label = label
+    label.items << self unless label.items.include?(self)
+  end
+
```
Comment (jerryowusu) [inline item.rb:17]: The item class should have one Genre, Author, and Label. As you have done for the Label on lines 13 - 16, kindly provide same for the Genre and Author. :wink:

**Review by jerryowusu — CHANGES_REQUESTED** [review]
Hi Team

Good job so far!
Your project is looking fantastic so far :star_struck: 

However, there are few changes you need to make in order to move on to the next level :sunglasses:

## Highlights
:heavy_check_mark: Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) is created in a separate .rb file
:heavy_check_mark: Most public methods have unit tests

## Required Changes ♻️

- [ ] As part of the project requirements, kindly ensure that your project documentation video contains a highlight about a piece of the code the you find interesting or is proud of making.
- [ ] Also do not forget that the maximum length of the video is 5 minutes. I would have ignored that but since you have to work on the video again kindly stick to the time constraints. :wink: 

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Zuhaib042 — APPROVED** [review]
Hi @peterdtitan @jerryowusu @preciousbetine 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✅ Professional Documentation
✅ Implemented suggested changes
✅ Linters passing
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**